File: nafed_erp/public/js/travel_request.js
frappe.ui.form.on("Travel Request", {
    employee: function(frm) {
        if (!frm.doc.employee) return;

        frappe.db.get_value("Employee", frm.doc.employee, "expense_approver")
            .then(r => {
                if (r && r.message && r.message.expense_approver) {
                    frm.set_value("custom_expense_approver", r.message.expense_approver);
                } else {
                    frm.set_value("custom_expense_approver", "");
                    frappe.msgprint("No Expense Approver set for this Employee.");
                }
            });
    },
    refresh(frm) {
        if (frm.doc.docstatus !== 1) return;

        // EMPLOYEE ADVANCE
        if (
            frm.doc.custom_payment_type === "Employee Advance" &&
            !frm.doc.custom_employee_advance &&
            frm.doc.custom_status === "Approved"
        ) {
            frm.add_custom_button(__("Create Employee Advance"), () => {
                frappe.call({
                    method: "nafed_erp.hire_to_retire.doc_events.travel_request.create_employee_advance_from_tr",
                    args: {
                        travel_request: frm.doc.name
                    },
                    callback(r) {
                        if (r.message) {
                            frappe.msgprint({
                                title: __("Employee Advance Created"),
                                message: __(
                                    "Employee Advance <b>{0}</b> has been successfully created for this Travel Request.",
                                    [r.message]
                                ),
                                indicator: "green"
                            });
        
                            frm.reload_doc();
                        }
                    }
                });
            });
        }
        

        // CREATE EXPENSE CLAIM (after Employee Advance exists)
        if (
            frm.doc.custom_payment_type === "Employee Advance" &&
            frm.doc.custom_employee_advance && frm.doc.custom_status==="Approved" && !frm.doc.custom_expense_claim
        ) {
            frm.add_custom_button(__("Create Expense Claim"), () => {
                frappe.call({
                    method: "nafed_erp.hire_to_retire.doc_events.travel_request.make_expense_claim_from_travel_request",
                    args: {
                        source_name: frm.doc.name
                    },
                    callback(r) {
                        if (r.message) {
                            frappe.model.sync(r.message);
                            frappe.set_route("Form", r.message.doctype, r.message.name);
                        }
                    }
                });
                
            });
        }
        // REIMBURSEMENT → EXPENSE CLAIM
        if (
            frm.doc.custom_payment_type === "Reimbursement" &&
            !frm.doc.custom_expense_claim && frm.doc.custom_status==="Approved"
        ) {
            frm.add_custom_button(__("Create Expense Claim"), () => {
                frappe.call({
                    method: "nafed_erp.hire_to_retire.doc_events.travel_request.make_expense_claim_from_travel_request",
                    args: {
                        source_name: frm.doc.name
                    },
                    callback(r) {
                        if (r.message) {
                            frappe.model.sync(r.message);
                            frappe.set_route("Form", r.message.doctype, r.message.name);
                        }
                    }
                });
                
            });
        }
    }
});

frappe.ui.form.on("Travel Itinerary", {
    departure_date(frm, cdt, cdn) {
        calculate_duration(frm, cdt, cdn);
    },

    arrival_date(frm, cdt, cdn) {
        calculate_duration(frm, cdt, cdn);
    }
});

function calculate_duration(frm, cdt, cdn) {
    let row = frappe.get_doc(cdt, cdn);

    if (row.departure_date && row.arrival_date) {
        let total_minutes = frappe.datetime.get_minute_diff(row.arrival_date, row.departure_date);

        let hours = Math.floor(total_minutes / 60);
        let minutes = total_minutes % 60;

        let formatted = `${hours}h ${minutes}m`;
        frappe.model.set_value(cdt, cdn, "custom_travel_duration", formatted);
    }
}


frappe.ui.form.on("Travel Request", {
    before_save: function(frm) {
        calculate_total_amount(frm);
    }
});

// ---------------------
// 2️⃣ Sum Total Amount of Costings Child Table
// ---------------------
function calculate_total_amount(frm) {
    let total = 0;

    (frm.doc.costings || []).forEach(row => {
        if (row.total_amount) {
            total += flt(row.total_amount);
        }
    });

    frm.set_value("custom_total_travel_amount", total);
}


frappe.ui.form.on("Travel Request", {
    before_save: function (frm) {
        let total_advance = 0;

        (frm.doc.itinerary || []).forEach(row => {
            total_advance += flt(row.advance_amount);
        });

        frm.set_value("custom_total_advance_amount", total_advance);
    }
});


frappe.ui.form.on("Travel Request Costing", {
    sponsored_amount: function (frm, cdt, cdn) {
        calculate_total(cdt, cdn);
    },

    funded_amount: function (frm, cdt, cdn) {
        calculate_total(cdt, cdn);
    }
});

function calculate_total(cdt, cdn) {
    let row = locals[cdt][cdn];

    let sponsored = row.sponsored_amount || 0;
    let funded = row.funded_amount || 0;

    row.total_amount = sponsored + funded;

    refresh_field("costings");
}

File: nafed_erp/nafed_erp/hire_to_retire/doc_events/travel_request.py
import re
import frappe
from frappe.utils import nowdate
from frappe.utils import add_years, getdate
from frappe.model.mapper import get_mapped_doc

def fill_submission_date(doc,method):
        # set the custom submission date
        doc.custom_submission_date = nowdate()

        # save the value in database
        doc.db_set("custom_submission_date", nowdate())


# Server Script: Travel Request (Event: Validate)

def parse_duration_to_minutes(duration):
    if not duration:
        return 0

    s = str(duration).strip().lower()
    s = s.replace(u"\u00A0", " ").replace("\t", " ").strip()

    # 24h 0m -> match hours and minutes
    hours = 0
    minutes = 0

    h = re.search(r"(\d+)\s*h", s)
    if h:
        hours = int(h.group(1))

    m = re.search(r"(\d+)\s*m", s)
    if m:
        minutes = int(m.group(1))

    return hours * 60 + minutes


def convert_minutes_to_duration(total_minutes):
    hours = total_minutes // 60
    minutes = total_minutes % 60
    return f"{hours}h {minutes}m"


def set_total_minutes(doc, method):
    total_minutes = 0

    for row in doc.itinerary:
        if row.custom_travel_duration:
            total_minutes += parse_duration_to_minutes(row.custom_travel_duration)

    doc.custom_total_travel_duration = convert_minutes_to_duration(total_minutes)

def validate_travel_request_ltc_period(doc, method):
    """
    Prevents an employee from creating more than one Travel Request
    within the LTC Period corresponding to the travel_category and posting_date.
    """

    if not doc.employee or not doc.custom_travel_category or not doc.custom_posting_date:
        return  # basic checks

    travel_date = getdate(doc.custom_posting_date)

    # 1️⃣ Get the LTC Period for this travel_category where the travel_date falls within the period
    ltc_period = frappe.get_all(
        "LTC Period",
        filters=[
            ["travel_category", "=", doc.custom_travel_category],
            ["period_start", "<=", travel_date],
            ["period_end", ">=", travel_date]
        ],
        fields=["name", "period_start", "period_end"],
        limit=1
    )

    if not ltc_period:
        frappe.throw(
            f"No LTC Period found for Travel Category '{doc.custom_travel_category}' "
            f"covering the posting date {travel_date}"
        )

    period_start = getdate(ltc_period[0].period_start)
    period_end = getdate(ltc_period[0].period_end)

    # 2️⃣ Check if employee already has a submitted Travel Request in this LTC period
    existing_tr = frappe.get_all(
        "Travel Request",
        filters=[
            ["employee", "=", doc.employee],
            ["custom_travel_category", "=", doc.custom_travel_category],
            ["docstatus", "=", 1],
            ["custom_posting_date", ">=", period_start],
            ["custom_posting_date", "<=", period_end]
        ],
        fields=["name"],
        limit=1
    )

    if existing_tr:
        frappe.throw(
            f"Employee {doc.employee} has already submitted a Travel Request "
            f"for '{doc.custom_travel_category}' in this LTC Period ({period_start} - {period_end})."
        )

def validate_advance_amount(doc,method): 
    if doc.custom_total_advance_amount > doc.custom_total_travel_amount: 
        frappe.throw("Toatal Advance Amount cannot be more than Total Amount.")

@frappe.whitelist()
def create_employee_advance_from_tr(travel_request):
    tr = frappe.get_doc("Travel Request", travel_request)

    # Extra safety: DB check
    existing = frappe.db.exists(
        "Employee Advance",
        {"custom_travel_request": tr.name, "docstatus": ["!=", 2]}
    )
    if existing:
        frappe.throw("Employee Advance already exists for this Travel Request.")

    ea = frappe.new_doc("Employee Advance")
    ea.employee = tr.employee
    ea.exchange_rate = 1  # VERY IMPORTANT
    ea.advance_amount = tr.custom_total_advance_amount
    ea.purpose = f"Travel Advance - {tr.name}"
    ea.custom_travel_request = tr.name
    ea.insert(ignore_permissions=True)
    ea.submit()

    # Save reference back to Travel Request
    frappe.db.set_value(
        "Travel Request",
        tr.name,
        "custom_employee_advance",
        ea.name
    )

    return ea.name


def cancel_linked_employee_advance(doc, method):
    print("Hello world")
    if not doc.custom_employee_advance:
        return

    ea = frappe.get_doc("Employee Advance", doc.custom_employee_advance)
    print("Cancelling employee advance!!")
    # Cancel Employee Advance first
    if ea.docstatus == 1:
        ea.cancel()

    # Clear reference in Travel Request (allowed in on_cancel)
    frappe.db.set_value(
        "Travel Request",
        doc.name,
        "custom_employee_advance",
        None
    )


@frappe.whitelist()
def make_expense_claim_from_travel_request(source_name, target_doc=None):

    def postprocess(source, target, source_parent):
        target.employee = source.employee
        target.custom_travel_request_ = source.name
        target.expense_approver = source.custom_expense_approver
        target.remarks = f"Travel reimbursement for {source.name}"

    def map_costings(source_row, target_row, source_parent):
        target_row.expense_type = source_row.expense_type
        target_row.amount = source_row.total_amount
        target_row.sanctioned_amount = source_row.total_amount

    return get_mapped_doc(
        "Travel Request",
        source_name,
        {
            "Travel Request": {
                "doctype": "Expense Claim",
                "postprocess": postprocess
            },
            "Travel Request Costing": {
                "doctype": "Expense Claim Detail",
                "postprocess": map_costings
            }
        },
        target_doc
    )